transformer_engine/pytorch/attention.py

```diff
@@ -116,6 +117,416 @@
+            "Supported: head_dim %%8 = 0, head_dim <= 256 (>192 requires sm80/90). "
+            "Found: head_dim = %s on sm%s.",
+            head_dim,
+            ".".join([str(i) for i in device_compute_capability]),
+        )
+        use_flash_attention = False
+
+    # Filter: QKV layout
+    qkv_format = "".join([i for i in qkv_layout.split("_")[0] if i.isalpha()])
+    if qkv_format == "thd":
+        if use_unfused_attention:
+            logger.debug("Disabling UnfusedDotProductAttention for qkv_format = thd")
+            use_unfused_attention = False
+        if use_flash_attention and pad_between_seqs:
+            logger.debug(
+                "Disabling FlashAttention for qkv_format = thd when there is "
+                "padding between sequences, i.e. [a, a, PAD, b, b, b, PAD, c, PAD]"
+            )
+            use_flash_attention = False
+
+    # Filter: Attention mask
+    # attn_mask_type               |     supported backends
+    # -------------------------------------------------------------------
+    # no_mask                      |     All
+    # padding                      |     FlashAttention, FusedAttention
+    # causal                       |
+    #     self-attention           |     All
+    #     cross-attention          |     FusedAttention
+    # padding_causal               |
+    #     self-attention           |     FlashAttention, FusedAttention
+    #     cross-attention          |     FusedAttention
+    # causal_bottom_right          |     All
+    # padding_causal_bottom_right  |     FlashAttention, FusedAttention
+    # arbitrary                    |     UnfusedDotProductAttention
+    if attn_mask_type == "arbitrary":
+        if use_flash_attention:
+            logger.debug("Disabling FlashAttention for arbitrary mask")
+        use_flash_attention = False
+        if use_fused_attention:
+            logger.debug("Disabling FusedAttention for arbitrary mask")
+        use_fused_attention = False
+    if use_unfused_attention and "padding" in attn_mask_type:
+        logger.debug("Disabling UnfusedDotProductAttention for padding masks")
+        use_unfused_attention = False
+    if use_unfused_attention and attn_mask_type == "causal" and max_seqlen_q != max_seqlen_kv:
+        logger.debug(
+            "Disabling UnfusedDotProductAttention for "
+            "top-left-diagonal causal masks for cross-attention"
+        )
+        use_unfused_attention = False
+    if (
+        use_flash_attention
+        and _flash_attn_2_1_plus
+        and attn_mask_type in ["causal", "padding_causal"]
+        and max_seqlen_q != max_seqlen_kv
+    ):
+        logger.warning(
+            "Disabling FlashAttention as it only supports bottom-right-diagonal "
+            "causal mask since flash-attn 2.1. See "
+            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
+        )
+        use_flash_attention = False
```

Comment:
```suggestion
    if (
        use_flash_attention
        and _flash_attn_2_1_plus
        and attn_mask_type in ["causal", "padding_causal"]
        and max_seqlen_q != max_seqlen_kv
    ):
        logger.warning(
            "Disabling FlashAttention as it only supports bottom-right-diagonal "
            "causal mask since flash-attn 2.1. See "
            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
        )
        use_flash_attention = False
    if (
        use_flash_attention
        and not _flash_attn_2_1_plus
        and attn_mask_type in ["causal_bottom_right", "padding_causal_bottom_right"]
        and max_seqlen_q != max_seqlen_kv
    ):
        logger.warning(
            "Disabling FlashAttention as it only supports top-left-diagonal "
            "causal mask before flash-attn 2.1. See "
            "https://github.com/Dao-AILab/flash-attention#21-change-behavior-of-causal-flag"
        )
        use_flash_attention = False
```